src/smtp_observer.py

```diff
@@ -0,0 +1,162 @@
+    """
+
+    _RELATION_NAME = "smtp"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "smtp-observer")
+        self._charm = charm
+        self.smtp = SmtpRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+        )
+        self.framework.observe(
+            self.smtp.on.smtp_data_available,
+            self._on_smtp_relation_data_available,
+        )
+
+    def get_relation_as_smtp_conf(self) -> Optional[SMTPConfiguration]:
+        """Get SMTP data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+        """
+        try:
+            relation_data: Optional[SmtpRelationData] = self.smtp.get_relation_data()
+        except ValidationError:
+            # ValidationError happens in the smtp(_legacy)relation_created event, as
+            # the relation databag is empty at that point.
+            logger.info("SMTP databag is empty. SMTP information will be set in the next event.")
+            return None
+
+        if relation_data is None:
+            return None
+
+        user = None
+        password = None
+        if relation_data.auth_type == AuthType.PLAIN:
+            user = relation_data.user
+            password = self._get_password_from_relation_data(relation_data)
+
+        # Not all combinations for the next variables are correct. See:
+        # https://github.com/matrix-org/synapse/blob/develop/synapse/config/emailconfig.py
+        force_tls = False
+        enable_tls = False
+        require_transport_security = False
+        if relation_data.transport_security == TransportSecurity.STARTTLS:
+            enable_tls = True
+            require_transport_security = True
+        elif relation_data.transport_security == TransportSecurity.TLS:
+            force_tls = True
+            enable_tls = True
+            require_transport_security = True
+
+        return SMTPConfiguration(
+            enable_tls=enable_tls,
+            force_tls=force_tls,
+            require_transport_security=require_transport_security,
+            host=relation_data.host,
+            port=relation_data.port,
+            user=user,
+            password=password,
+        )
+
+    def _get_password_from_relation_data(self, relation_data: SmtpRelationData) -> Optional[str]:
+        """Get smtp password from relation data.
+
+        Arguments:
+            relation_data: The relation data from where to extract the password
+
+        Returns:
+            the password or None if no password found
+        """
+        # If the relation data password_id exists, that means that
+        # Juju version is >= 3.0 and secrets are used for the password.
+        # Otherwise, use the field password as a fallback
+        if relation_data.password_id:
```

Comment:
You can use the JUJU_HAS_SECRETS variable to make it clear.

Reply:
I thought about that, however in the smtp-integrator charm, there are [two relations](https://github.com/canonical/smtp-integrator-operator/blob/main/metadata.yaml), `smtp` and `smtp-legacy`. `smtp` works only if juju has secrets, but `smtp-legacy` works in all cases. Using juju secrets will break `smtp-legacy`. WDYT?

Reply:
Actually I thougt about just replacing "if relation_data.password_id" for "if JUJU_HAS_SECRETS". Wouldnt that work?

Reply:
Until  [PR](https://github.com/canonical/smtp-integrator-operator/pull/29) smtp-integrator with secrets was not working correctly.

It now works, so I have updated the tests so that now it runs tests for `smtp` and `smtp-legacy` endpoints and updated the integration tests to run in Juju 3.

If only JUJU_HAS_SECRETS is checked, then this will not work in Juju 3: `juju integrate smtp-integrator:smtp-legacy synapse:smtp`, as secrets are not used in that endpoint in the smtp-integrator charm, so the test will fail. I suppose this could be used for example for offers.

However, now that smtp integrator works with secrets, not checking the data bag fields and instead checking with just JUJU_HAS_SECRETS is possible, but then it will not be possible to use `smtp-legacy` in Juju 3. I do not have a strong opinion, do you prefer to drop the support for `smtp-legacy`?

Reply:
Lets support both for now since the staging env isn't Juju 3 yet.